test/unit/Cardano/Wallet/AddressDerivationSpec.hs

```diff
@@ -0,0 +1,234 @@
+                "Ae2tdPwUPEZLSqQN7XNJRMJ6yHWdfFLaQgPPYgyJKrJnCVnRtbfw6EHRv1D"
+
+
+{-------------------------------------------------------------------------------
+                               Properties
+-------------------------------------------------------------------------------}
+
+
+prop_succMaxBoundHardIx :: Property
+prop_succMaxBoundHardIx = expectFailure $
+    property $ succ (maxBound @(Index 'Hardened _)) `seq` ()
+
+prop_predMinBoundHardIx :: Property
+prop_predMinBoundHardIx = expectFailure $
+    property $ pred (minBound @(Index 'Hardened _)) `seq` ()
+
+prop_succMaxBoundSoftIx :: Property
+prop_succMaxBoundSoftIx = expectFailure $
+    property $ succ (maxBound @(Index 'Soft _)) `seq` ()
+
+prop_predMinBoundSoftIx :: Property
+prop_predMinBoundSoftIx = expectFailure $
+    property $ pred (minBound @(Index 'Soft _)) `seq` ()
+
+prop_toEnumChangeChain :: Int -> Property
+prop_toEnumChangeChain n =
+    n > fromEnum InternalChain ==> expectFailure $ property $
+        (toEnum n :: ChangeChain) `seq` ()
+
+prop_roundtripEnumChangeChain :: ChangeChain -> Property
+prop_roundtripEnumChangeChain ix =
+    (toEnum . fromEnum) ix === ix
+
+prop_roundtripEnumIndexHard :: Index 'Hardened 'AccountK -> Property
+prop_roundtripEnumIndexHard ix =
+    (toEnum . fromEnum) ix === ix .&&. (toEnum . fromEnum . getIndex) ix === ix
+
+prop_roundtripEnumIndexSoft :: Index 'Soft 'AddressK -> Property
+prop_roundtripEnumIndexSoft ix =
+    (toEnum . fromEnum) ix === ix .&&. (toEnum . fromEnum . getIndex) ix === ix
+
+-- | Deriving address public key should be equal to deriving address
+-- private key and extracting public key from it (works only for non-hardened
+-- child keys).
+--
+-- To compute the public child key of a parent private key:
+--  * N(CKDpriv((kpar, cpar), i)) (works always).
+--  * CKDpub(N(kpar, cpar), i) (works only for non-hardened child keys).
+--
+-- Thus:
+--
+-- N(CKDpriv((kpar, cpar), i)) === CKDpub(N(kpar, cpar), i)
+--
+-- if (kpar, cpar) is a non-hardened key.
+--
+-- For details see <https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki#private-parent-key--public-child-key bip-0039>
+prop_publicChildKeyDerivation
+    :: (Seed, Passphrase "generation")
+    -> Passphrase "encryption"
+    -> ChangeChain
+    -> Index 'Soft 'AddressK
+    -> Property
+prop_publicChildKeyDerivation (Seed seed, recPwd) encPwd cc ix =
+    addrXPub1 === addrXPub2
+  where
+    accXPrv = unsafeGenerateKeyFromSeed (seed, recPwd) (encPwd)
+    -- N(CKDpriv((kpar, cpar), i))
+    addrXPub1 = publicKey $ deriveAddressPrivateKey encPwd accXPrv cc ix
+    -- CKDpub(N(kpar, cpar), i)
+    addrXPub2 = deriveAddressPublicKey (publicKey accXPrv) cc ix
+
+prop_accountKeyDerivation
+    :: (Seed, Passphrase "generation")
+    -> Passphrase "encryption"
+    -> Index 'Hardened 'AccountK
+    -> Property
+prop_accountKeyDerivation (Seed seed, recPwd) encPwd ix =
+    accXPub `seq` property ()
```

Comment:
Is this checking for absence of `error`?

Reply:
Yes. Making sure that this code can run just fine.